### PR TITLE
[Enhancement] Improve equality checks in layout nodes and fragment validation

### PR DESCRIPTION
**Pull request by LeiWang1999**
## Summary

This PR improves the correctness of layout and fragment equality checks, and adds validation for physical index compatibility in `ProveFragmentContains`.

### Problem

1. **`LayoutNode::IsEqual` and `FragmentNode::IsEqual` had incorrect comparison logic**: The original implementation used `StructuralEqual` directly on `forward_index_`, which could incorrectly compare `InputPlaceholder` variables. For example, two layouts with different mappings like `i * 16 + j` and `j * 16 + i` might be incorrectly considered equal if their `InputPlaceholder` indices happened to match structurally.

2. **`ProveFragmentContains` only checked thread containment, not physical index compatibility**: When validating a loop layout candidate against a buffer fragment, the function only verified that threads accessing small fragment elements are a subset of threads accessing large fragment elements. However, it didn't verify that the physical indices match, which could lead to incorrect code generation.

   **Example of incorrect code generation:**
   ```
   Loop layout:   (i, j) -> physical = [0],  thread = i*4+j
   Buffer layout: (i, j) -> physical = [j],  thread = i*4+rep
   ```
   
   Without physical index validation, code generation would produce:
   ```cpp
   // Incorrect: col_sum[(threadIdx.x) & 3] 
   // Should be: col_sum[j] where j is correctly derived
   sums[(threadIdx.x) + 128] = col_sum[(threadIdx.x) & 3];
   ```

### Solution

1. **Fix `IsEqual` by using `Forward` with common variables**:
   ```cpp
   // Create common variables for comparison
   Array<PrimExpr> common_vars;
   for (size_t i = 0; i < this->InputDim(); i++) {
     common_vars.push_back(Var("_cmp_v" + std::to_string(i)));
   }
   
   // Compare Forward results with StructuralEqual
   auto this_forward = this->Forward(common_vars);
   auto other_forward = other->Forward(common_vars);
   if (!StructuralEqual()(this_forward, other_forward)) {
     return false;
   }
   ```

2. **Add `check_forward_index` parameter to `ProveFragmentContains`**:
   ```cpp
   bool ProveFragmentContains(Fragment small_frag, Fragment large_frag,
                              Array<PrimExpr> small_frag_indices,
                              Array<PrimExpr> large_frag_indices,
                              arith::Analyzer &analyzer_,
                              bool check_forward_index = false);
   ```
   
   When `check_forward_index=true`, validates physical index equality:
   ```cpp
   auto small_physical = small_frag->Forward(small_frag_indices);
   auto large_physical = large_frag->Forward(large_frag_indices);
   // Dimension mismatch or value mismatch -> return false
   ```

### Changes

- `src/op/parallel.h`: Add `check_forward_index` parameter to `ProveFragmentContains`
- `src/op/parallel.cc`: Implement physical index validation; update call sites in `ValidateCandidateAgainstFragments` and `InferLayout`
- `src/layout/layout.cc`: Fix `LayoutNode::IsEqual` and `FragmentNode::IsEqual` to use `Forward` with common variables
- `testing/python/layout/test_tilelang_layout_equal.py`: Add comprehensive tests for layout/fragment equality
- Move `test_tilelang_issue_layout.py` to `testing/python/layout/test_tilelang_layout_inference.py`

## Test plan

- [x] All 24 new layout equality tests pass
- [x] Existing layout inference test passes
- [x] Manual verification that incorrect code generation case is now caught

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved layout equality validation to be more robust against structural differences in internal representations.
  * Enhanced fragment containment verification with forward index validation to ensure correctness during layout operations.

* **Tests**
  * Added comprehensive test suite for layout and fragment equality semantics, covering edge cases and complex scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by LeiWang1999** [comment]
@codex review

**Comment by LeiWang1999** [comment]
@regression-perf

File: src/op/parallel.cc [hunk 1]
```diff
@@ -44,7 +44,30 @@ using namespace tir;
...
+  // When check_forward_index is true, verify that the physical indices
+  // (forward index) of both fragments are equal. This is required when
+  // validating loop layout against buffer fragment, as code generation
+  // needs to correctly derive buffer physical indices from loop layout.
+  if (check_forward_index) {
+    auto small_physical = small_frag->Forward(small_frag_indices);
+    auto large_physical = large_frag->Forward(large_frag_indices);
+
+    // Dimension mismatch means they are not equal.
+    if (small_physical.size() != large_physical.size()) {
```
[the remark this answers is not in the document]
Comment (LeiWang1999) [reply src/op/parallel.cc:58]: Thanks for the review! However, this concern doesn't apply here.

  When a Fragment is constructed with an empty forward_index (e.g., Fragment(..., {}, ...)), the constructor automatically infers and fills it via infer_fragment_index():

```
  // layout.cc:510-513
  if (forward_index.empty()) {
    forward_index = {
        infer_fragment_index(getVarMap(), forward_thread_, &analyzer)};
  }
```
  So after construction, all Fragments have non-empty forward_index_. The "empty forward_index" scenario you described doesn't exist at runtime - it's only a construction-time shorthand that gets filled automatically.
